README.md
# videojs-scene7

[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![Build Status](https://github.com/amclin/videojs-scene7/actions/workflows/run-tests.yml/badge.svg)](https://github.com/amclin/videojs-scene7/actions/workflows/run-tests.yml)
[![Coverage Status][codecov-icon]][codecov-link]
[![David Dependancy Status](https://david-dm.org/amclin/videojs-scene7.svg)](https://david-dm.org/amclin/videojs-scene7)
[![David Dev Dependancy Status](https://david-dm.org/amclin/videojs-scene7/dev-status.svg)](https://david-dm.org/amclin/videojs-scene7?type=dev)

[![NPM][npm-icon]][npm-link]

Tech plugin for VideoJS to support Scene7 players
## Installation

```sh
npm install --save videojs-scene7
```

## Usage

To include videojs-scene7 on your website or web application, use any of the following methods.

### `<script>` Tag with a `<video>` tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available. Make sure to set:
- techOrder so VideoJS will leverage the Scene7 plugin
- source of the Scene7 MediaSet
- the type to `videojs/scene7`
- The Scene7 [player options](#options)

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/videojs-scene7.min.js"></script>
<video id='my-video'
  controls
  autoplay
  >
  <source src='Scene7SharedAssets/Adobe_QBP-AVS' type='videojs/scene7'>
</video>
<script>
  var player = videojs('my-video',{
    techOrder: [ 'Scene7']
  });

  player.scene7();
</script>
```

### Browserify/CommonJS

When using with Browserify, install videojs-scene7 via npm and `require` the plugin as you would any other module.

```js
var videojs = require('video.js');

// The actual plugin function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require('videojs-scene7');

var player = videojs('my-video',{
  techOrder: [ 'Scene7']
});

player.scene7();
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(['video.js', 'videojs-scene7'], function(videojs) {
  var player = videojs('my-video',{
    techOrder: [ 'Scene7']
  });

  player.scene7();
});
```

## Options
Provide Scene7 options to VideoJS through the normal VideoJS options object or data-options attribute. The options

```js
options: {
  scene7: {
    serverurl: "https://s7d1.scene7.com/is/image/", // Path to Scene7 server
    videoserverurl: "https://s7d1.scene7.com/is/content/", // 
  }
}
```

### `serverurl`

URL to the image API for Scene7. Default when not set is `https://s7d1.scene7.com/is/image/`

### `videoserverurl`

URL to the content API for Scene7. Default when not set is `https://s7d1.scene7.com/is/content/`

### `contenturl`

URL to the content API for Scene7 where losed caption and chapter navigation assets are located. Default when not set is `https://s7d1.scene7.com/is/content/`


## License

MIT. Copyright (c) Anthony McLin &lt;npm@anthonymclin.com&gt;


[videojs]: https://videojs.com/

[codecov-icon]: https://codecov.io/gh/amclin/videojs-scene7/branch/master/graph/badge.svg?token=qT4GevGPZO

[codecov-link]: https://codecov.io/gh/amclin/videojs-scene7

[npm-icon]: https://nodei.co/npm/videojs-scene7.png?downloads=true&downloadRank=true

[npm-link]: https://nodei.co/npm/videojs-scene7/


CHANGELOG.md
## [4.0.3](https://github.com/amclin/videojs-scene7/compare/v4.0.2...v4.0.3) (2023-02-10)


### Bug Fixes

* **deps:** bump @xmldom/xmldom from 0.7.6 to 0.7.9 ([a03fff7](https://github.com/amclin/videojs-scene7/commit/a03fff7bc1db4a0f2c53b89c208a7ff242332568))
* **deps:** bump minimist from 1.2.5 to 1.2.6 ([b81a6c5](https://github.com/amclin/videojs-scene7/commit/b81a6c57166b03ae7ce5104e6ddb1a6b9be6a53e))
* **deps:** bump ua-parser-js from 0.7.31 to 0.7.33 ([56c1a55](https://github.com/amclin/videojs-scene7/commit/56c1a55b243df70d6b05853c980b1c6e58a52699))

## [4.0.2](https://github.com/amclin/videojs-scene7/compare/v4.0.1...v4.0.2) (2023-02-10)


### Bug Fixes

* rename config files to support Rollup 3 ([cc61f86](https://github.com/amclin/videojs-scene7/commit/cc61f86380479eb07dbac1785af10cb07279c706))

## [4.0.1](https://github.com/amclin/videojs-scene7/compare/v4.0.0...v4.0.1) (2023-02-10)


### Bug Fixes

* **deps:** bump minimist from 1.2.5 to 1.2.6 ([e1d3b45](https://github.com/amclin/videojs-scene7/commit/e1d3b453fbb6f3bd0b15fbcaf8905430c29f88d0))

# [4.0.0](https://github.com/amclin/videojs-scene7/compare/v3.0.0...v4.0.0) (2023-02-10)


### Features

* **security:** switch all HTTP to HTTPS ([5a89645](https://github.com/amclin/videojs-scene7/commit/5a896456b2c887f08c79257ae513a62020933070)), closes [#1034](https://github.com/amclin/videojs-scene7/issues/1034)


### BREAKING CHANGES

* **security:** defaults are now https instead of http

Connections to Adobe-provided scripts are now done
using HTTPS instead of HTTP. Should not be an issue

# [3.0.0](https://github.com/amclin/videojs-scene7/compare/v2.2.11...v3.0.0) (2023-02-10)


### Bug Fixes

* **deps:** bump http-cache-semantics from 4.1.0 to 4.1.1 ([9bb2687](https://github.com/amclin/videojs-scene7/commit/9bb2687079eaacb51a54f52885d2e32dd5735b4c))
* **deps:** bump ua-parser-js from 0.7.31 to 0.7.33 ([f2f9791](https://github.com/amclin/videojs-scene7/commit/f2f9791ff6b7a71254984890f891a443a31719b1))


### Continuous Integration

* start testing Node 18 future-proofed using `lts` keyword ([ddbda81](https://github.com/amclin/videojs-scene7/commit/ddbda810fff3ca3c26f3c6fba25dbb9f71138664))


### Features

* support Node 18 and future LTS releases ([3ed1008](https://github.com/amclin/videojs-scene7/commit/3ed10084ad4d0aa7203d1c3ba7b906b31006839c))


### BREAKING CHANGES

* no longer supports Node 12

Various libraries are starting to drop support for Node 12.
While this only affects the dev dependencies of this project
it's only a matter of time before something critical drops
support. So to keep maintenance easy, we're dropping
v12 now as it is past EOL already and moving to 14, 16,
and 18 LTS versions.

Recommended action: upgrade your project from Node 12
to a later LTS release (14, 16, 18)
* no longer supports Node 12

Various libraries are starting to drop support for Node 12.
While this only affects the dev dependencies of this project
it's only a matter of time before something critical drops
support. So to keep maintenance easy, we're dropping
v12 now as it is past EOL already and moving to 14, 16,
and 18 LTS versions.

Recommended action: upgrade your project from Node 12
to a later LTS release (14, 16, 18)

## [2.2.11](https://github.com/amclin/videojs-scene7/compare/v2.2.10...v2.2.11) (2023-01-04)


### Bug Fixes

* **deps:** bump @xmldom/xmldom from 0.7.6 to 0.7.9 ([5744616](https://github.com/amclin/videojs-scene7/commit/5744616095e7e4558bc7c4d68ca91bb0c707f3db))

## [2.2.10](https://github.com/amclin/videojs-scene7/compare/v2.2.9...v2.2.10) (2022-11-22)


### Bug Fixes

* **deps:** bump engine.io from 6.2.0 to 6.2.1 ([2229ef7](https://github.com/amclin/videojs-scene7/commit/2229ef756a949b2643e1bcf1d57a49ab0337737a))

## [2.2.9](https://github.com/amclin/videojs-scene7/compare/v2.2.8...v2.2.9) (2022-11-15)


### Bug Fixes

* **deps:** bump minimatch from 3.0.4 to 3.1.2 ([8f5cccd](https://github.com/amclin/videojs-scene7/commit/8f5cccdbd303c8feb2c789d0f0cef84d6aa64efb))

## [2.2.8](https://github.com/amclin/videojs-scene7/compare/v2.2.7...v2.2.8) (2022-10-25)


### Bug Fixes

* **deps:** bump @xmldom/xmldom from 0.7.5 to 0.7.6 ([3454d85](https://github.com/amclin/videojs-scene7/commit/3454d85003c4994acc43eea488b6c56882ff32da))

## [2.2.7](https://github.com/amclin/videojs-scene7/compare/v2.2.6...v2.2.7) (2022-06-22)


### Bug Fixes


test/s7faker.js
/**
 * Scene7 API Faker
 *
 * simulates the Scene7 API for use within unit tests
 * because it takes to long to load the real libraries
 * from Scene7
 */

import document from 'global/document';

const events = {};
const params = {
  params: {},

  init() {
    return;
  },

  push(arg1, arg2) {
    params.params[arg1] = arg2;
  }
};
const mediaset = {
  component: { }
};
const player = {
  defaults: {
    volume: 0.5,
    currentTime: 5000,
    duration: 100000,
    loadedPosition: 75000,
    width: 200,
    height: 100,
    isMuted: false
  },

  play() {
    return;
  },

  pause() {
    return;
  },

  resize(width, height) {
    player.defaults.width = width;
    player.defaults.height = height;
  },

  seek(param) {
    player.defaults.currentTime = param;
  },

  getCurrentTime() {
    return player.defaults.currentTime;
  },

  getDuration() {
    return player.defaults.duration;
  },

  getHeight() {
    return player.defaults.height;
  },

  getLoadedPosition() {
    return player.defaults.loadedPosition;
  },

  supportsVolumeControl() {
    return true;
  },

  getVolume() {
    return player.defaults.volume;
  },

  setVolume(vol) {
    player.defaults.volume = vol;
  },

  mute(m) {
    if (m) {
      player.defaults.isMuted = true;
    }
  },

  unmute() {
    player.defaults.isMuted = false;
  },

  muted() {
    return player.defaults.isMuted;
  },

  getWidth() {
    return player.defaults.width;
  }
};
const container = {
  defaults: {
    width: 400,
    height: 200
  },
  component: { },
  resize(width, height) {
    container.defaults.width = width;
    container.defaults.height = height;
  },
  getHeight() {
    return player.defaults.height;
  },
  getWidth() {
    return player.defaults.width;
  }
};
const s7faker = {
  // Fake the loader
  Util: {
    lib: {
      include() {
        return true;
      }
    },

    init() {
      return true;
    }
  },

  Event: {
    SDK_READY: 'SDK_READY'
  },

  event: {
    AssetEvent: {
      NOTF_SET_PARSED: 'NOTF_SET_PARSED'
    },
    CapabilityStateEvent: {
      NOTF_VIDEO_CAPABILITY_STATE: 'NOTF_VIDEO_CAPABILITY_STATE'
    },
    ResizeEvent: {
      FULLSCREEN_RESIZE: 'FULLSCREEN_RESIZE',
      COMPONENT_RESIZE: 'COMPONENT_RESIZE'
    },
    StatusEvent: {
      NOTF_VIEW_READY: 'NOTF_VIEW_READY'
    },
    UserEvent: {
      PAUSE: 'PAUSE',
      PLAY: 'PLAY'
    },
    VideoEvent: {
      NOTF_CURRENT_TIME: 'NOTF_CURRENT_TIME',
      NOTF_DURATION: 'NOTF_DURATION',
      NOTF_VIDEO_END: 'NOTF_VIDEO_END',
      NOTF_VOLUME: 'NOTF_VOLUME'
    }
  },

  ParameterManager() {
    return params;
  },

  common: { },

  set: {
    MediaSet() {
      mediaset.component = {
        mediaSet_: params.params['MediaSet.asset'], // eslint-disable-line camelcase
        settings: {
          params: params.params
        }
      };
      return mediaset;
    }
  },

  video: {
    VideoPlayer() {
      return player;
    }
  }
};

/**
 * Simulates Scene7's addEventListener
 *
 * @param {string} name The event to listen to
 * @param {function} callback A method to trigger when the event happens
 */
function addEventListener(name, callback) {
  if (typeof events[name] === 'undefined') {
    events[name] = [];
  }

  if (typeof callback === 'function') {
    events[name].push(callback);
  }
}

/**
 * Simulates Scene7's dispatchEvent method
 *
 * @param {string} name The event to trigger
 */
function dispatchEvent(name) {
  events[name].forEach((callback) => {
    const ev = {
      s7event: {
        w: 100,
        h: 100
      }
    };

    callback(ev);
  });
}

/**
 * Simulates Scene7's addEventListener
 *
 * @param {Object} arg Not implemented. Unclear what this does, in S7 examples it's generally null
 * @param {Object} opts A Scene7 ParameterManager containing configuration
 * @param {string} id Used to set the ID on the <div> created for the container
 * @return {Object} Returns a Scene7 container element
 */
function initContainer(arg, opts, id) {
  let node;

  if (typeof container.component.obj === 'undefined') {
    node = document.createElement('div');
    node.id = id;
    container.component.obj = node;
    document.body.appendChild(node);
  }

  return container;
}

mediaset.addEventListener = addEventListener;
mediaset.dispatchEvent = dispatchEvent;
params.addEventListener = addEventListener;
params.dispatchEvent = dispatchEvent;
player.addEventListener = addEventListener;
player.dispatchEvent = dispatchEvent;
container.addEventListener = addEventListener;
container.dispatchEvent = dispatchEvent;
s7faker.common.Container = initContainer;
container.requestFullScreen = function() {
  container.dispatchEvent(s7faker.event.ResizeEvent.FULLSCREEN_RESIZE);
};
container.cancelFullScreen = function() {
  container.dispatchEvent(s7faker.event.ResizeEvent.COMPONENT_RESIZE);
};

export default s7faker;


src/plugin.js
import videojs from 'video.js';
import {version as VERSION} from '../package.json';
import window from 'global/window';

const Tech = videojs.getComponent('Tech');

// Default options for the plugin.
const defaults = {
  serverurl: 'https://s7d1.scene7.com/is/image/',
  videoserverurl: 'https://s7d1.scene7.com/is/content/',
  // specify content url for closed caption and chapter navigation asset
  contenturl: 'https://s7d1.scene7.com/is/content/',
  // configures the icon effect when video is in paused state
  // as follows: enable,# times to appear, fade duration, auto-hide duration
  iconeffect: '0,-1,0.3,0'
};

// Scene7 tracks at a different timescale than VideoJS
const timeRatio = 1000;

// Cross-compatibility for Video.js 5 and 6.
// const registerTech = videojs.registerTech || videojs.plugin;
// const dom = videojs.dom || videojs;

/**
 * Scene7 Media Controller - Wrapper for Scene7 Player API
 *
 * @extends Tech
 */
class Scene7 extends Tech {

  /**
  * Create an instance of this Tech.
  *
  * @param {Object} [options]
  *        The key/value store of player options.
  *
  * @param {Component~ReadyCallback} ready
  *        Callback function to call when the `Scene7` Tech is ready.
  */
  constructor(options, ready) {
    // Merge the videojs options with the defaults
    options = Object.assign({}, defaults, options);

    super(options, ready);

    this.s7 = {};

    this._loadS7SDK();
    this._setupS7Params(options);
  }

  /**
  * Load and Initialize the Scene7 SDK
  */
  _loadS7SDK() {

    // TODO add the SDK path from defaults

    // S7 SDK is in the global namespace
    this.s7.sdk = window.s7sdk;

    const sdk = this.s7.sdk;

    // Load specific modules
    sdk.Util.lib.include('s7sdk.common.Container');
    sdk.Util.lib.include('s7sdk.set.MediaSet');
    sdk.Util.lib.include('s7sdk.video.VideoPlayer');

    // Initialize the SDK
    sdk.Util.init();
  }

  /**
   * Create a S7 Parameters manager
   */
  _setupS7Params() {
    const sdk = this.s7.sdk;
    const that = this;
    /* Create an instance of the ParameterManager component to collect
    components' configuration that can come from a viewer preset, URL,
    or the HTML page itself. The ParameterManager component also sends
    a notification s7sdk.Event.SDK_READY when all needed files are
    loaded and the configuration parameters are processed. The other
    components should never be initialized outside of this handler.
    After defining the handler for the s7sdk.Event.SDK_READY event, it
    is safe to initiate configuration initialization by calling
    ParameterManager.init(). */
    const params = new sdk.ParameterManager();

    /* Setup event to call next steps when PackageManager dispatches
    s7sdk.Event.SDK_READY event when all modifiers are processed and
    it is safe to initialize the viewer. */
    params.addEventListener(that.s7.sdk.Event.SDK_READY, function(event) {
      that._initViewer();
    }, false);

    // Initialize ParameterManager
    params.init();

    this.s7.params = params;
  }

  /**
   * Maps videojs options to Scene7 parameters
   * and sets them in the ParameterManager object
   **/
  setParameters() {
    const paramMgr = this.s7.params;
    const options = this.options_;
    const params = {};
    // List of supported settings to get from the videojs options
    const settings = [
      'serverurl',
      'videoserverurl',
      'contenturl',
      'loop',
      'iconeffect'
    ];

    // Get supported settings from videojs
    settings.forEach(function(setting) {
      if (typeof options[setting] !== 'undefined') {
        params[setting] = options[setting];
      }
    });

    // Add source Media Set from <video> <source> tag
    params['MediaSet.asset'] = options.source.src;

    // Convert autoplay setting from boolean to numeric strings
    params.autoplay = (this.options_.autoplay) ? '1' : '0';

    // Convert loop setting from boolean to numeric strings
    params.loop = (this.options_.loop) ? '1' : '0';

    // Provide settings to Scene7 ParametersManager
    for (const param in params) {
      paramMgr.push(param, params[param]);
    }

  }

  /**
   * Initialize the various parts of the Scene7 Video Player
   */
  _initViewer() {
    this.setParameters();
    this._setupS7Container();
    this._setupS7MediaSet();
    this._setupS7Player();
    this._mapEvents();
    this.injectS7Player();
    this.triggerReady();
  }

  /**
   * Create the Scene7 Adaptive Video media set
   *
   * @param {string} src
   *    - Optional media set path to use. Replaces value in ParamsManager when provided.
   */
  _setupS7MediaSet(src) {
    const that = this;
    const sdk = that.s7.sdk;
    const container = that.s7.container;
    const params = that.s7.params;
    let mediaSet = {};

    // Update parameters if specified
    if (src) {
      params.params['MediaSet.asset'] = src;
    }

    // Get new MediaSet object
    mediaSet = new sdk.set.MediaSet(container, params, container.parentId + '-s7mediaSet');

    // Add MediaSet event listeners
    mediaSet.addEventListener(sdk.event.AssetEvent.NOTF_SET_PARSED, function(event) {
      that._setS7Source(event.s7event.asset);
    }, false);

    // Store new MediaSet
    this.s7.mediaSet = mediaSet;
  }

  /**
   * Creates the Scene7 Viewer Container as a parent for the
   * other S7 UI components that are part of the application
   **/
  _setupS7Container() {
    const that = this;
    const sdk = that.s7.sdk;
    const params = that.s7.params;

    that.el_.id = that.el_.id || that.getUniqueId();

    const container = new sdk.common.Container(that.el_.id, params, that.el_.id + '-s7container');

    // Setup events to resize the player when the container changes size/fullscreen
    container.addEventListener(sdk.event.ResizeEvent.COMPONENT_RESIZE, function(event) {
      that._resizeEventHandler(event);
    }, false);
    container.addEventListener(sdk.event.ResizeEvent.FULLSCREEN_RESIZE, function(event) {
      that._resizeEventHandler(event);
    }, false);

    that.s7.container = container;
  }

  /**
   * Handles Scene7 resize events and triggers a resize of the video player
   *
   * @param {Object} event
   *        Scene7 Event object
   *
   **/
  _resizeEventHandler(event) {
    const width = event.s7event.w;
    const height = event.s7event.h;

    this.resizeVideo(width, height);
  }

  /**
   * Creates the Scene7 Video Player component used inside the container
   */
  _setupS7Player() {
    const that = this;
    const sdk = that.s7.sdk;
    const container = that.s7.container;
    const params = that.s7.params;
    const player = new sdk.video.VideoPlayer(container, params, container.parentId + '-s7viewer');

    that.s7.player = player;
  }

  /**
   * Set the Scene7 source for the player
   *
   * @param {Object} asset
   *        Scene7 Media Set object
   *
   **/
  _setS7Source(asset) {
    const that = this;
    const sdk = that.s7.sdk;
    const player = that.s7.player;
    let src = {};

    // Ensure this is a media set asset
    if (!(asset instanceof sdk.MediaSetDesc)) {
      throw new Error('Failed to get meta data for video: ' + asset);
    }

    if (asset.type === sdk.ItemDescType.VIDEO_SET) {
      // Adaptive Video Set
      src = asset;
    } else {
      // Single Video
      src = asset.items[0];
    }

    player.setItem(src);
  }

  /**
   * A getter/setter for the `Html5` Tech's source object.
   * > Note: Please use {@link Html5#setSource}
   *
   * @param {Tech~SourceObject} [src]
   *        The source object you want to set on the `HTML5` techs element.
   *
   * @return {Tech~SourceObject|undefined}
   *         - The current source object when a source is not passed in.
   *         - undefined when setting
   *
   * @deprecated Since version 5.
   */
  src(src) {
    if (typeof src === 'undefined') {
      return this.currentSrc();
    }

    // Setting src through `src` instead of `setSrc` will be deprecated
    this.setSrc(src);
  }

  /**
   * Set the video source for Scene7
   * {@link Tech~SourceObject} for the media.
   *
   * @method Scene7#setSrc
   * @param {Tech~SourceObject} src
   *        The source object to set as the current source.
   *
   * @see [Spec]{@link https://www.w3.org/TR/html5/embedded-content-0.html#dom-media-src}
   */
  setSrc(src) {
    this._setupS7MediaSet(src.src);
  }

  /**
   * Get the current source on the Scene7 Tech.
   *
   * @return {Tech~SourceObject}
   *         The current source object from the Scene7 tech.
   */
  currentSrc() {
    return this.options_.source;
  }

  /**
   * Starts video playback. Restarts the video if we're already at the end
   *
   * @return {undefined}
   **/
  play() {
    const that = this;
    const player = that.s7.player;

    // IF the video is over, restart from the beginning
    if (that.ended()) {
      that.setCurrentTime(0);
    }

    return player.play();
  }

  /**
   * Checks if the video has reached the end or not
   *
   * @return {boolean}
   *    - True if ended
   *    - False if not ended
   **/
  ended() {
    const player = this.s7.player;

    if (this.state === 'ended') {
      return true;
    }

    return (player.getDuration() - player.getCurrentTime() <= 1);
  }

  /**
   * Sets the current playback point
   *
   * @param {number} time
   *    - Time in seconds since the beginning
   */
  setCurrentTime(time) {
    this.s7.player.seek(time * timeRatio);
    super.setCurrentTime();
  }

  /**
   * Pauses video playback
   *
   * @return {undefined}
   **/
  pause() {
    this.s7.player.pause();
  }

  /**
   * Get the value of `paused` from the Scene7. `paused` indicates whether the
   * media element is currently paused or not.
   *
   * @method Scene7#paused
   * @return {boolean}
   *         Whether or not Scene7 is paused
   *
   * @see [Spec]{@link https://www.w3.org/TR/html5/embedded-content-0.html#dom-media-paused}
   */
  paused() {
    // We track a state managed by events since Scene7 internal
    // player.resolveVideoProxy().paused() is unreliable and always
    // returns true
    // We pass 'ended' as paused since VideoJS expects this when a
    // video is ended and pending replay.
    return this.state === 'paused' || this.state === 'ended';
  }

  /**
   * Get the current playback time
   *
   * @return {number}
   *    - seconds since beginning
   **/
  currentTime() {
    return this.s7.player.getCurrentTime() / timeRatio;
  }

  /**
   * Check if the volume can be changed in this browser/device.
   * Volume cannot be changed in a lot of mobile devices.
   * Specifically, it can't be changed from 1 on iOS.
   *
   * @return {boolean}
   *         - True if volume can be controlled
   *         - False otherwise
   */
  canControlVolume() {
    return this.s7.player.supportsVolumeControl();
  }

  /**
   * Get the current volume level
   *
   * @return {number}
   *    The current volume level as a fractional value
   *    0 = min
   *    1 = max
   **/
  volume() {
    return this.s7.player.getVolume();
  }

  /**
   * Set the value of `volume` on the media element. `volume` indicates the current
   * audio level as a percentage in decimal form. This means that 1 is 100%,
   * 0.5 is 50%, and * so on.
   *
   * @method Scene7#setVolume
   * @param {number} percentAsDecimal
   *        The volume percent as a decimal. Valid range is from 0-1.
   * @return {undefined} on success
   *
   * @see [Spec]{@link https://www.w3.org/TR/html5/embedded-content-0.html#dom-a-volume}
   */
  setVolume(percentAsDecimal) {
    return this.s7.player.setVolume(percentAsDecimal);
  }

  /**
   * Check if video is muted
   *
   * @return {boolean} muted
   *  - True when video is muted
   *  - False when video is not muted
   * @method muted
   */
  muted() {
    return this.s7.player.muted();
  }

  /**
   * Mute/Unmute the video
   *
   * @method Scene7#setMuted
   * @param {boolean} muted
   *        - True if the audio should be set to silent
   *        - False otherwise
   * @return {undefined} on success
   */
  setMuted(muted) {
    const player = this.s7.player;

    // ensure boolean
    muted = Boolean(muted);

    if (muted) {
      return player.mute(muted);
    }

    return player.unmute();
  }

  /**
   * Get the current duration of the Scene7 media element.
   *
   * @return {number}
   *         The duration of the media in seconds or NaN if there is no duration.
   */
  duration() {
    return this.s7.player.getDuration() / timeRatio;
    // TODO Duration may not be available until playback has started
    // If that's the case, then use a this.on('timeupdate',checkProgress)
    // listener to poll for a duration update.
    // see https://github.com/videojs/video.js/blob/master/src/js/tech/html5.js
  }

  /**
   * Get the value of `buffered` from the media element. `buffered` is a `TimeRange`
   * object that represents the parts of the media that are already downloaded and
   * available for playback.
   *
   * @return {TimeRange}
   *         The value of `buffered` from the media element.
   *
   * @see [Spec]{@link https://www.w3.org/TR/html5/embedded-content-0.html#dom-media-buffered}
   */
  buffered() {
    const that = this;
    const player = that.player;
    let timeRange = {
      length: 0,
      start() {
        throw new Error('This TimeRanges object is empty');
      },
      end() {
        throw new Error('This TimeRanges object is empty');
      }
    };

    if (player && player.getLoadedPosition) {
      timeRange = {
        length: 1,
        start() {
          return 0;
        },
        end() {
          return (player.getLoadedPosition() * player.getDuration()) / timeRatio;
        }
      };
    } else {
      that.trigger('error');
    }

    return timeRange;
  }

  /**
   * Check if fullscreen is supported on the current playback device.
   *
   * @return {boolean}
   *         - True if fullscreen is supported.
   *         - False if fullscreen is not supported.
   */
  supportsFullScreen() {
    // Scene7 handles full-screen mode internally depending on browser type
    return true;
  }

  /**
   * Requests that Scene7 Tech open full screen mode.
   * will have different behaviors on different device types
   *
   * @return {undefined}
   **/
  enterFullScreen() {
    return this.s7.container.requestFullScreen();
  }

  /**
   * Requests that Scene7 Tech exits full screen mode.
   *
   * @return {undefined}
   **/
  exitFullScreen() {
    return this.s7.container.cancelFullScreen();
  }

  /**
   * Map required Video JS events to their Scene7 equivalents
   **/
  _mapEvents() {
    const that = this;
    const sdk = that.s7.sdk;
    const player = that.s7.player;
    const events = {
      player: [
        {
          internal: 'UserEvent.PLAY',
          external: 'play'
        }, {
          internal: 'UserEvent.PAUSE',
          external: 'pause'
        }, {
          internal: 'StatusEvent.NOTF_VIEW_READY',
          external: 'playing'
        }, {
          internal: 'VideoEvent.NOTF_CURRENT_TIME',
          external: 'timeupdate'
        }, {
          internal: 'VideoEvent.NOTF_DURATION',
          external: 'durationchange'
        }, {
          internal: 'VideoEvent.NOTF_VIDEO_END',
          external: 'ended'
        }, {
          internal: 'VideoEvent.NOTF_VOLUME',
          external: 'volumechange'
        }
      ],
      container: [
        {
          internal: 'ResizeEvent.FULLSCREEN_RESIZE',
          external: 'enterFullScreen'
        }
      ]
    };

    // wire the S7 event to the Video JS using an anonymous function
    for (const component in events) {
      events[component].forEach(function(event) {
        const splitName = event.internal.split('.');
        const target = that.s7[component];
        const s7Event = sdk.event[splitName[0]][splitName[1]];
        const vjsEvent = event.external;

        target.addEventListener(s7Event, function(ev) {
          that.trigger(vjsEvent);
        }, false);
      });
    }

    // Events that require some logic when mapping
    player.addEventListener(
      sdk.event.VideoEvent.NOTF_LOAD_PROGRESS,
      function(event) {
        that._handleLoadProgress();
      },
      false
    );

    // Video is switched into playable/pausable/replayable state
    player.addEventListener(
      sdk.event.CapabilityStateEvent.NOTF_VIDEO_CAPABILITY_STATE,
      function(event) {
        that._handleStateChange(event);
      },
      false
    );
  }

  /**
   * Deals with the logic of switching which events to fire
   * loadstart fires only once, but progress and duration change
   * get updated multiple times
   **/
  _handleLoadProgress() {
    const that = this;

    // Only trigger loadstart once
    if (that.loadstartFired) {
      that.trigger('loadedmetadata');
      that.trigger('loadstart');
      that.loadstartFired = true;
    }

    that.trigger('progress');
    that.trigger('durationchange');
  }

  /**
   * Tracks the Scene7 player state since some methods are not exposed
   *
   * @param {Object} event
   *      Scene7 event
   */
  _handleStateChange(event) {
    const sdk = this.s7.sdk;
    const cap = event.s7event.state;

    // Scene7 is pausable means the video is playing
    if (cap.hasCapability(sdk.VideoCapabilityState.PAUSE)) {
      // this.trigger('timeupdate');
      // this.trigger('durationchange');
      this.trigger('playing');
      this.trigger('play');
      this.state = 'playing';
      return;
    }

    // Scene7 is playable means the video is paused
    if (cap.hasCapability(sdk.VideoCapabilityState.PLAY)) {
      this.trigger('canplay');
      this.trigger('pause');
      this.state = 'paused';
      return;
    }

    // if (cap.hasCapability(sdk.VideoCapabilityState.STOP)) {
    //   self.state = 'playing';
    //   return;
    // }

    // Scene7 is replayable means the video is ended
    if (cap.hasCapability(sdk.VideoCapabilityState.REPLAY)) {
      this.trigger('canplay');
      this.trigger('ended');
      this.state = 'ended';
      return;
    }
  }

  /**
   * Resize the video when the container resizes
   *
   * @param {number} width
   *    - Width of the video in pixels
   * @param {number} height
   *    - Height of the video in pixels
   **/
  resizeVideo(width, height) {
    this.s7.player.resize(width, height);
  }

  /**
   * Resize the container
   *
   * @param {number} width
   *    - Width of the container in pixels
   * @param {number} height
   *    - Height of the container in pixels
   **/
  resizeContainer(width, height) {
    this.s7.container.resize(width, height);
  }

  /**
   * Move the Scene7 container inside the VideoJS container
   */
  injectS7Player() {
    const that = this;
    const s7El = that.s7.container.component.obj;
    const vjsEl = that.el_.parentElement;

    that.el_.appendChild(s7El);

    that.resizeContainer(vjsEl.offsetWidth, vjsEl.offsetHeight);
    that.resizeVideo(vjsEl.offsetWidth, vjsEl.offsetHeight);
  }

  /**
   * Genterate a unique id
   * https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
   *
   * @return {string} A unique ID in the format x-xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
   */
  getUniqueId() {
    return 'x-' + 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);

      return v.toString(16);
    });
  }
}

// ****** These methods fail to exist if defined in the class ****** //

/**
 * Check if Scene7 video is supported by this browser/device
 * (this doesn't work as a native function)
 *
 * @return {boolean} Always true when S7 is enabled
 */
Scene7.isSupported = function() {
  return true;
};

/**
 * Check if the tech can support the given mime/type
 *
 * @param {string} mimetype
 *        mime/type of the video
 *
 * @return {string}
 *    - 'probably' if S7 can play the video
 *    - 'maybe'
 *    - '' empty string if S7 cannot play the video
 **/
Scene7.canPlayType = function(mimetype) {
  return mimetype === 'videojs/scene7' ? 'probably' : '';
};

/**
 * Check if the tech can support the given source
 *
 * @param  {Object} srcObj  The source object
 *
 * @return {string}
 *    - 'probably' if S7 can play the video
 *    - 'maybe'
 *    - '' empty string if S7 cannot play the video
 */
Scene7.canPlaySource = function(srcObj) {
  return this.canPlayType(srcObj.type);
};

/**
 * Set the tech's volume control support status
 *
 * @type {Boolean}
 */
Scene7.featuresVolumeControl = true;

// Register the plugin with video.js.
videojs.registerTech('Scene7', Scene7);

// Include the version number.
Scene7.VERSION = VERSION;

export default Scene7;
